### PR TITLE
Adding new focus trap options support

### PR DESCRIPTION
**Pull request by Dan503**
- Adding 2 new animated demos
- Adding the new values to PropTypes
- Adding support for `checkCanReturnFocus` and `onPostDeactivate`
- Adding Cypress tests for the new demos

<details>
<summary>PR Checklist</summary>
<br/>

__Please leave this checklist in your PR.__

- Issue being fixed is referenced.
- Source changes maintain:
  - Stated browser compatibility.
  - Stated React compatibility.
- Unit test coverage added/updated.
- E2E test coverage added/updated.
- Prop-types added/updated.
- Typings added/updated.
- README updated (API changes, instructions, etc.).
- Changes to dependencies explained.
- Changeset added (run `yarn changeset` locally to add one, and follow the prompts).
  - EXCEPTION: A Changeset is not required if the change does not affect any of the source files that produce the package bundle. For example, tooling changes, test updates, or a new dev-only dependency to run tests more efficiently should not have a Changeset since it will not affect package consumers.

</details>

### CODE REVIEW

File: demo/js/demo-animated-dialog.js
Comment (Dan503) [inline demo/js/demo-animated-dialog.js:35]: I know that you are going to ask me to move this to the CSS style sheet. I want to keep this here though.

Unlike with the core `focus-trap` demos, the react demos are able to show everything relevant to recreating the demo in the one file nice the HTML is also stored in this JS file.

The styles for the animated dialogs are an important aspect of the demo so they should be displayed as part of the demo and not hidden in the main css file.

File: demo/js/demo-animated-trigger.js
Comment (Dan503) [inline demo/js/demo-animated-trigger.js:35]: Same as above, the styles for the animated demos are an important part of the demo so should be displayed as part of the demo.

File: src/focus-trap-react.js
Comment (Dan503) [inline src/focus-trap-react.js:69]: I'm not sure if this should be called regardless of weather `this.previouslyFocusedElement` exists or if it should only be called if focus is able to be returned.

I leaned more toward the "only call if focus is returned" option.

File: src/focus-trap-react.js
Comment (Dan503) [inline src/focus-trap-react.js:200]: Can these prop type options be moved into the core focus-trap library instead of having to replicate them in all the spin-off focus-trap projects?

It will make maintaining them much easier.

File: cypress/integration/focus-trap-demo.spec.js
Comment (Dan503) [inline cypress/integration/focus-trap-demo.spec.js:70]: Animated demo tests are just a copy-paste of the default test.

They don't test for a console log message after `onPostActivate` and `onPostDeactivate` are called.

**Review by Dan503 — COMMENTED** [review]
Some additional things for you to be aware of while reviewing.

**Comment by stefcameron** [comment]
👀

File: src/focus-trap-react.js
Comment (stefcameron) [reply src/focus-trap-react.js:67]: Actually, I think we got this slightly wrong over in focus-trap:

https://github.com/focus-trap/focus-trap/blob/master/index.js#L548-L557
```javascript
      const finishDeactivation = () => {
        if (returnFocus) {
          delay(() => {
            tryFocus(getReturnFocusNode(state.nodeFocusedBeforeActivation));
            if (onPostDeactivate) {
              onPostDeactivate();
            }
          });
        }
      };
```

I think there's a bug there in that `onPostDeactivate()` is only called if focus is returned. Doesn't matter if focus is returned or not, the trap still got deactivated. The code should probably be this:

```javascript
      const finishDeactivation = () => {
        delay(() => {
          if (returnFocus) {
            tryFocus(getReturnFocusNode(state.nodeFocusedBeforeActivation));
          }
          if (onPostDeactivate) {
            onPostDeactivate();
          });
        }
      };
```

Meaning that onPostDeactivate() would consistently be called whenever the trap is deactivated, whether focus is returned or not, since the trap is always deactivated regardless of focus being returned...

And then it would inform this decision here: To always call it even if focus isn't returned.

👉 I can create a small PR for this over on focus-trap if you agree with my thinking.

One way or another, however, it means that because the point at which focus is returned is altered by focus-trap-react, we need to also alter the point at which onPostDeactivate() is called so it matches the focus-trap behavior.

That means that when we're building the `this.tailoredFocusTrapOptions` (see https://github.com/focus-trap/focus-trap-react/blob/master/src/focus-trap-react.js#L34), we need to set grab `props.focusTrapOptions.onPostDeactivate` and keep it in a special class member, like `this.originalPostDeactivate` and always set `onPostDeactivate: null` in `this.tailoredFocusTrapOptions` (which is what we ultimately use to create the trap at https://github.com/focus-trap/focus-trap-react/blob/master/src/focus-trap-react.js#L77)

And then it means that here, we should not do this:
```javascript
    const { checkCanReturnFocus, onPostDeactivate } = this.props.focusTrapOptions;
```
but rather this:
```javascript
    const { checkCanReturnFocus } = this.tailoredFocusTrapOptions;
    const onPostDeactivate = this.originalPostDeactivate;
```

Makes sense?

Finally, we also need to call onPostDeactivate() here if `this.returnFocusOnDeactivate` is false:
- https://github.com/focus-trap/focus-trap-react/blob/master/src/focus-trap-react.js#L109
- https://github.com/focus-trap/focus-trap-react/blob/master/src/focus-trap-react.js#L134

We should probably change the internal `returnFocus()` method to `deactivateTrap()` which does this:
```javascript
if (this.focusTrap) {
  // NOTE: we never let the trap return the focus since we do that ourselves
  this.focusTrap.deactivate({ returnFocus: false });
}

if (this.returnFocusOnDeactivate && this.previouslyFocusedElement && this.previouslyFocusedElement.focus) {
    this.previouslyFocusedElement.focus();
  }
}

if (this.originalPostDeactivate) {
  this.originalPostDeactivate.call(null); // don't call it in context of "this"
}
```

Then we can call `deactivateTrap()` from all affected locations where the trap needs to be deactivated:
- https://github.com/focus-trap/focus-trap-react/blob/master/src/focus-trap-react.js#L105
- https://github.com/focus-trap/focus-trap-react/blob/master/src/focus-trap-react.js#L129

**Review by stefcameron — CHANGES_REQUESTED** [review]
Thank you, @Dan503 ! Big help here too. I added some thoughts, LMK what you think!

File: src/focus-trap-react.js
Comment (Dan503) [reply src/focus-trap-react.js:67]: I'm a bit confused about the `this.tailoredFocusTrapOptions` stuff but it makes sense that you want to call `postDeactivate` regardless of weather focus is sent since the function name doesn't really say anything about returning focus. Maybe we got the name wrong. Maybe we should have called it `onReturnFocus`.

I had a look and it looks like `onPostActivate` is called regardless of weather the focus is sent or not so `onPostDeactivate` should be consistent.

Maybe we provide both `onPostDeactivate` and `onReturnFocus` options? Though I'm not sure if there is enough value in doing that to justify having the extra callback option. Probably leave it out for now.

File: src/focus-trap-react.js
Comment (Dan503) [reply src/focus-trap-react.js:198]: Ok, yeah PropTypes is more of a React thing. 

It just means that it is not as simple as just installing a new focus-trap version when the settings change.

By spin-off projects I was thinking there would also be Vue, Angular, Svelt etc. versions of the project.

File: cypress/integration/focus-trap-demo.spec.js
Comment (Dan503) [reply cypress/integration/focus-trap-demo.spec.js:70]: In order to increase the code coverage I might need to make the tests slightly different 😕
Though there is definitely some refactoring that can be done here. I was hoping to not have to put to much effort into test writing though.

Or do you think the coverage is enough as is? You can see the code coverage warnings in the PR.

I should probably add tests for checking that the source code links go to the right place.

File: src/focus-trap-react.js
Comment (stefcameron) [reply src/focus-trap-react.js:67]: > I'm a bit confused about the `this.tailoredFocusTrapOptions` stuff [...]

It's because of the fact that focus-trap-react does some special handling around focus-trap's `returnFocusOnDeactivate ` because of focus issues that come-up when React is in charge of rendering. focus-trap-react sets its own flag and kind of replicates the behavior in the place of focus-trap. And because onPostDeactivate() in focus-trap comes _after_ focus is returned, we now need to specially handle that option too, so that focus-trap doesn't call onPostDeactivate() before focus-trap-react has a chance to return the focus, since the trap will actually get deactivated _before_ focus-trap-react finishes its processing. Hopefully that's a bit clearer.

> I had a look and it looks like `onPostActivate` is called regardless of weather the focus is sent or not so `onPostDeactivate` should be consistent.

OK, I will make a small PR to focus-trap to fix this.

> Maybe we provide both `onPostDeactivate` and `onReturnFocus` options? Though I'm not sure if there is enough value in doing that to justify having the extra callback option. Probably leave it out for now.

I say leave that out for now. We can add that later if there's a real need for it.

File: cypress/integration/focus-trap-demo.spec.js
Comment (stefcameron) [reply cypress/integration/focus-trap-demo.spec.js:70]: I know, I loathe tests. 😄 But they help give us confidence. It would be good to at least cover the fact that the new options are being called as expected. That should be fairly easy to do as unit tests in https://github.com/focus-trap/focus-trap-react/blob/master/test/focus-trap-react.test.js I don't think the Cypress tests need to add more coverage.

> I should probably add tests for checking that the source code links go to the right place.

If you mean the "view source" links, I don't think that's necessary. It's more important to spend time on testing actual library functionality.

Comment (Dan503) [reply cypress/integration/focus-trap-demo.spec.js]: Ok, I've added unit tests now and made sure that they fail if the options are not behaving correctly.

I didn't add unit tests to core focus-trap though when I added the new options. You might want to take the test I wrote and apply them to core focus-trap.

Any reason you don't have code coverage set up for core focus-trap?

**Comment by Dan503** [comment]
@stefcameron ready for you to take another look at now.

File: src/focus-trap-react.js
Comment (Dan503) [inline src/focus-trap-react.js:148]: With the pause stuff, `this.focusTrap.activate()` has some asynchronous functionality now if someone has defined `checkCanActivate`.

Do you think that will that be problematic?

File: cypress/integration/focus-trap-demo.spec.js
Comment (stefcameron) [reply cypress/integration/focus-trap-demo.spec.js]: > Any reason you don't have code coverage set up for core focus-trap?

Because the person who set it up for this repo didn't for that one. 🙂 🤷‍♂️ More of a nice-to-have for me right now.

**Comment by stefcameron** [comment]
PR posted on focus-trap to fix the onPostDeactivate issue: https://github.com/focus-trap/focus-trap/pull/433

**Comment by stefcameron** [comment]
I will check out your changes here a bit later.

File: test/focus-trap-react.test.js
Comment (stefcameron) [inline test/focus-trap-react.test.js:333]: Is there a reason it has to be this long? Wouldn't just 5 or 10ms suffice?

File: test/focus-trap-react.test.js
Comment (stefcameron) [inline test/focus-trap-react.test.js:357]: Can this be combined into a `waitFor` statement? [waitFor](https://testing-library.com/docs/dom-testing-library/api-async#waitfor)'s default timeout is 1000ms so it should be more than enough.

```
waitFor(() => expect(hasActivated).toBe(true))
```

That would be more "in line" with testing with RTL.

File: test/focus-trap-react.test.js
Comment (stefcameron) [inline test/focus-trap-react.test.js:362]: Same here

File: test/focus-trap-react.test.js
Comment (stefcameron) [inline test/focus-trap-react.test.js:372]: Can this just be 5 or 10ms?

File: test/focus-trap-react.test.js
Comment (stefcameron) [inline test/focus-trap-react.test.js:409]: Same thing here about `waitFor` instead of using a timer.

File: test/focus-trap-react.test.js
Comment (stefcameron) [inline test/focus-trap-react.test.js:329]: Can we use mock functions instead these flags?
```
const onActivate = jest.fn();
const onPostActivate = jest.fn();
```
set these as the options, and then later,
```
expect(onActivate).toHaveBeenCalled();
```

File: test/focus-trap-react.test.js
Comment (stefcameron) [inline test/focus-trap-react.test.js:368]: Same here about using mock functions instead of the flags.

File: test/focus-trap-react.test.js
Comment (stefcameron) [inline test/focus-trap-react.test.js:415]: Same here for the mock functions...

File: test/focus-trap-react.test.js
Comment (stefcameron) [inline test/focus-trap-react.test.js:443]: Same here for the mock functions...

File: src/focus-trap-react.js
Comment (stefcameron) [inline src/focus-trap-react.js:77]: I think we need to do something similar to https://github.com/focus-trap/focus-trap/blob/master/index.js#L222 here, because we're ignoring the `setReturnFocus` option (as we were before; this is an existing bug).

File: src/focus-trap-react.js
Comment (stefcameron) [inline src/focus-trap-react.js:86]: Instead of just `this.previouslyFocusedElement`, it should look more like https://github.com/focus-trap/focus-trap/blob/master/index.js#L561 so that `setReturnFocus` option is handled properly since it won't be handled by focus-trap now.

**Review by stefcameron — CHANGES_REQUESTED** [review]
Thanks for adding the unit tests and making those changes to the source (and you thought `checkCanDeactivate`, which I hadn't considered, nice!).

File: src/focus-trap-react.js
Comment (stefcameron) [reply src/focus-trap-react.js:148]: Unpausing doesn't cause the trap to be re-activated by going through the activation sequence and calling onActivate, etc. Am I missing something?

File: test/focus-trap-react.test.js
Comment (Dan503) [reply test/focus-trap-react.test.js:357]: `waitFor` was the first thing I tried. `waitFor` has very different functionality though.

`waitFor` actually works like this:
1. Try calling the callback function immediately.
2. If it errors try again _immediately_
3. Keep trying over and over until it either stops throwing errors or the timeout time has been reached.

That is very different functionality to "wait x milliseconds" which is what the `pause()` function does.

File: test/focus-trap-react.test.js
Comment (Dan503) [reply test/focus-trap-react.test.js:333]: 5ms is probably OK...

File: test/focus-trap-react.test.js
Comment (Dan503) [reply test/focus-trap-react.test.js:329]: Oh, I didn't know about mock functions. Yeah that will probably work as long as there is also `expect(onActivate).toNotHaveBeenCalled()`

File: src/focus-trap-react.js
Comment (Dan503) [reply src/focus-trap-react.js:148]: The unit tests are passing...

Do you think you can make a failing unit test for this functionality?

File: test/focus-trap-react.test.js
Comment (stefcameron) [reply test/focus-trap-react.test.js:329]: Yes, but it should be `expect(onActivate).not.toHaveBeenCalled()`

File: test/focus-trap-react.test.js
Comment (stefcameron) [reply test/focus-trap-react.test.js:357]: You're right, they are different, but why wouldn't `waitFor()` work? The assertion will fail when `waitFor()` does the immediate call, which is fine. Then 5ms will elapse and onActivate will get called, and then 45ms later, `waitFor()` will again try the assertion, and it should pass at that point. No?

File: test/focus-trap-react.test.js
Comment (stefcameron) [reply test/focus-trap-react.test.js:333]: That would good. I'd like to avoid introducing unnecessary test time duration as much as possible. So let's use the shortest delays possible.

File: src/focus-trap-react.js
Comment (stefcameron) [reply src/focus-trap-react.js:148]: I don't think so. With the tests passing, I think we're good.

File: test/focus-trap-react.test.js
Comment (Dan503) [reply test/focus-trap-react.test.js:357]: I need to be able to easily check that the function has been called when it is supposed to have be called. 

Importantly I also need to check that the function has _not_ been called when it should not have been called yet.

I need to:
check `false` > **wait** > check `true`.

`waitFor` is a loop that just does this:
check `true` > check `true` > check `true`

Comment (stefcameron) [reply test/focus-trap-react.test.js:357]: Isn't that expected, though? Because focus-trap calls onActivate immediately on activation, then looks at checkCanFocusTrap, and then calls onPostActivate. 🤔

Comment (Dan503) [reply test/focus-trap-react.test.js:357]: This is a test. We want to test that the expected behaviour actually does what we expect it to do.

We don't want to make a change that causes `onPostActivate` to get called before `checkCanActivate` is resolved. We want to ensure that if that situation were to happen, our tests will fail so we can fix the bug before it gets released to the public.

Comment (stefcameron) [reply test/focus-trap-react.test.js:357]: OK, there must be something I'm not understanding. You're closer to the test than I am, so I will trust your judgment. Thanks for the discussion; let's go with using pause() as you're doing it now. :)

File: demo/js/demo-setReturnFocus.js
Comment (Dan503) [inline demo/js/demo-setReturnFocus.js:9]: > I think we need to do something similar to https://github.com/focus-trap/focus-trap/blob/master/index.js#L222 here, because we're ignoring the setReturnFocus option (as we were before; this is an existing bug).

I've added support for the `setReturnFocus` prop though the code coverage has dipped a bit since I don't think the entire `getNodeForOption` function is being used.

**Review by Dan503 — COMMENTED** [review]
Ready for you to take another look at now @stefcameron :)

File: src/focus-trap-react.js
Comment (stefcameron) [inline src/focus-trap-react.js:61]: Why are these here? None of these are options that are typed as `HTMLElement | SVGElement | string | () => HTMLElement | SVGElement` which is what we're interested in here.

File: test/focus-trap-react.test.js
Comment (stefcameron) [inline test/focus-trap-react.test.js:500]: Thanks for adding this!

**Review by stefcameron — COMMENTED** [review]
This is looking tight now! Just one question remaining.

File: src/focus-trap-react.js
Comment (Dan503) [reply src/focus-trap-react.js:61]: I just grabbed that straight from core `focus-trap`.

I didn't look much into it. I just assumed that they were necessary since it was in core.

https://github.com/focus-trap/focus-trap/blob/81898771081e524847c8c4dc3a942cc5f6340378/index.js#L90

I see that it is used for setting up default settings in core focus-trap. Not necessary here.

It's late for me. I'll need to fix it up tomorrow.

File: test/focus-trap-react.test.js
Comment (Dan503) [reply test/focus-trap-react.test.js:500]: Kind of didn't have much choice with the code coverage restriction on PRs for this repo

File: src/focus-trap-react.js
Comment (Dan503) [inline src/focus-trap-react.js:65]: Actually if you can replace `config` here with `this.tailoredFocusTrapOptions` then it will unblock the PR.

Comment (stefcameron) [reply src/focus-trap-react.js]: Done!

**Review by stefcameron — APPROVED** [review]
I say this is good to go! 😄

**Comment by Dan503** [comment]
Code coverage is blocking it.

Maybe just comment out the lines that coverage is complaining about. 

If we aren't testing them it means we aren't using them (yet).

If all the demo's seen in core focus trap are converted into React demo's here then we would be able to get to full coverage.

File: .changeset/quick-months-notice.md
Comment (Dan503) [inline .changeset/quick-months-notice.md:4]: What about mentioning `setReturnFocus` as well?

**Comment by stefcameron** [comment]
> Code coverage is blocking it.
> 
> Maybe just comment out the lines that coverage is complaining about.
> 
> If we aren't testing them it means we aren't using them (yet).
> 

I'll leave them as-is and take the slight hit on coverage for now. At least it's the same code in focus-trap that has all the coverage there.

> If all the demo's seen in core focus trap are converted into React demo's here then we would be able to get to full coverage.

Yes, that's true.

**Comment by Dan503** [comment]
> Maybe just comment out the lines that coverage is complaining about.

Alternatively you can write tests for them if you know how to go about doing that to increase the coverage but I'd just comment out those lines for now.

File: .changeset/quick-months-notice.md
Comment (stefcameron) [reply .changeset/quick-months-notice.md:4]: Missed this before I merged. Yes, that's a good idea. I'll add it to the changeset, thank you!

**Comment by stefcameron** [comment]
@all-contributors add @Dan503 for code, test

**Comment by stefcameron** [comment]
Being published in 8.6.0!

**Pull request merged**
